tools/swiftdt/main.swift

```diff
@@ -0,0 +1,166 @@
+    let protoName = context.name(proto: proto) ?? "<unknown>"
+    print("Conformance: \(typeName): \(protoName)")
+  })
+}
+
+func dumpMetadataAllocations(context: SwiftReflectionContextRef) throws {
+  var allocations: [swift_metadata_allocation_t] = []
+  var metadatas: [swift_reflection_ptr_t] = []
+  try context.iterateMetadataAllocations(call: { allocation in
+    allocations.append(allocation)
+    print("Metadata allocation at: \(hex: allocation.Ptr) " + 
+          "size: \(allocation.Size) tag: \(allocation.Tag)")
+    let ptr = context.metadataPointer(allocation: allocation)
+    if ptr != 0 {
+      metadatas.append(ptr)
+    }
+  })
+
+  allocations.sort(by: { $0.Ptr < $1.Ptr })
+  for metadata in metadatas {
+    let name = context.name(metadata: metadata) ?? "<unknown>"
+    print("Metadata \(hex: metadata)")
+    print("    Name: \(name)")
+
+    if let allocation = allocations.last(where: { metadata >= $0.Ptr }) {
+      let offset = metadata - allocation.Ptr
+      print("    In allocation \(hex: allocation.Ptr) " +
+            "size \(allocation.Size) at offset \(offset)")
+    } else {
+      print("    Not in any known metadata allocation. How strange.")
+    }
+  }
+}
+
+func printUsage(args: inout ArraySlice<String>) {
+  print("Usage: \(executableName) <command>", to: &Std.err)
+  print("", to: &Std.err)
+  print("Available commands:", to: &Std.err)
+  
+  let maxWidth = commands.map({ $0.name.count }).max() ?? 0
+  for command in commands {
+    let paddedName = command.name.padding(toLength: maxWidth,
+                                          withPad: " ", startingAt: 0)
+    print("  \(paddedName) - \(command.help)", to: &Std.err)
+  }
+}
+
+func makeReflectionContext(args: inout ArraySlice<String>)
+  -> (Inspector, SwiftReflectionContextRef) {
+  guard let pidStr = args.popFirst() else {
+    argFail("Must specify a pid or process name")
+  }
+
+  guard let pid = pidFromHint(pidStr) else {
+    argFail("Cannot find pid/process \(pidStr)")
+  }
+
+  guard let inspector = Inspector(pid: pid) else {
+    argFail("Failed to inspect pid \(pid) (are you running as root?)")
+  }
+
+  guard let reflectionContext
+    = swift_reflection_createReflectionContextWithDataLayout(
+    inspector.passContext(),
+    Inspector.Callbacks.QueryDataLayout,
+    Inspector.Callbacks.Free,
+    Inspector.Callbacks.ReadBytes,
+    Inspector.Callbacks.GetStringLength,
+    Inspector.Callbacks.GetSymbolAddress) else {
+    argFail("Failed to create reflection context")
+  }
+  return (inspector, reflectionContext)
+}
+
+func withReflectionContext(
+  args: inout ArraySlice<String>,
+  call: (SwiftReflectionContextRef) throws -> Void) throws {
+  let (inspector, context) = makeReflectionContext(args: &args)
+  defer {
+    swift_reflection_destroyReflectionContext(context)
+    inspector.destroyContext()
+  }
+  try call(context)
+}
+
+let commandName = argv.popFirst()
+for command in commands {
```

Comment:
Check that argv is now empty here, or else allow multiple commands to be run in one execution.